xds/src/main/java/io/grpc/xds/XdsClientImpl2.java

```diff
@@ -0,0 +1,1688 @@
+          // fall through
+        case ADS_TYPE_URL_RDS_V2:
+          return RDS;
+        case ADS_TYPE_URL_CDS:
+          // fall through
+        case ADS_TYPE_URL_CDS_V2:
+          return CDS;
+        case ADS_TYPE_URL_EDS:
+          // fall through
+        case ADS_TYPE_URL_EDS_V2:
+          return EDS;
+        default:
+          return UNKNOWN;
+      }
+    }
+  }
+
+  /**
+   * Tracks a single subscribed resource.
+   */
+  private final class ResourceSubscriber {
+    private final ResourceType type;
+    private final String resource;
+    private final Set<ResourceWatcher> watchers = new HashSet<>();
+    private ResourceUpdate data;
+    private boolean absent;
+    private ScheduledHandle respTimer;
+
+    ResourceSubscriber(ResourceType type, String resource) {
+      this.type = type;
+      this.resource = resource;
+      if (rpcRetryTimer != null && rpcRetryTimer.isPending()) {
+        return;
+      }
+      restartTimer();
+    }
+
+    void addWatcher(ResourceWatcher watcher) {
+      checkArgument(!watchers.contains(watcher), "watcher %s already registered", watcher);
+      watchers.add(watcher);
+      if (data != null) {
+        notifyWatcher(watcher, data);
+      } else if (absent) {
+        watcher.onResourceDoesNotExist(resource);
+      }
+    }
+
+    void removeWatcher(ResourceWatcher watcher) {
+      checkArgument(watchers.contains(watcher), "watcher %s not registered", watcher);
+      watchers.remove(watcher);
+    }
+
+    void restartTimer() {
+      class ResourceNotFound implements Runnable {
+        @Override
+        public void run() {
+          logger.log(XdsLogLevel.INFO, "{0} resource {1} initial fetch timeout", type, resource);
+          respTimer = null;
+          onAbsent();
+        }
+
+        @Override
+        public String toString() {
+          return type + this.getClass().getSimpleName();
+        }
+      }
+
+      respTimer = syncContext.schedule(
+          new ResourceNotFound(), INITIAL_RESOURCE_FETCH_TIMEOUT_SEC, TimeUnit.SECONDS,
+          timeService);
+    }
+
+    void stopTimer() {
+      if (respTimer != null && respTimer.isPending()) {
+        respTimer.cancel();
+        respTimer = null;
+      }
+    }
+
+    boolean isWatched() {
+      return !watchers.isEmpty();
+    }
+
+    void onData(ResourceUpdate data) {
+      if (respTimer != null && respTimer.isPending()) {
+        respTimer.cancel();
+        respTimer = null;
+      }
+      ResourceUpdate oldData = this.data;
+      this.data = data;
+      absent = false;
+      if (!Objects.equals(oldData, data)) {
+        for (ResourceWatcher watcher : watchers) {
+          notifyWatcher(watcher, data);
+        }
+      }
+    }
+
+    void onAbsent() {
+      if (respTimer != null && respTimer.isPending()) {  // too early to conclude absence
+        return;
+      }
+      logger.log(XdsLogLevel.INFO, "Conclude {0} resource {1} not exist", type, resource);
+      boolean oldAbsent = absent;
```

Comment:
Same comment as https://github.com/grpc/grpc-java/pull/7458#discussion_r496167917

Reply:
Oops, sorry sorry. I was working on this PR in parallel with the previous one, forgot to fix the same thing in this one.